beacon-chain/types/crystallized_state.go

```diff
@@ -0,0 +1,347 @@
+	return c.data.JustifiedStreak
+}
+
+// CrosslinkingStartShard returns next shard that crosslinking assignment will start from.
+func (c *CrystallizedState) CrosslinkingStartShard() uint64 {
+	return c.data.CrosslinkingStartShard
+}
+
+// LastJustifiedSlot return the last justified slot of the beacon chain.
+func (c *CrystallizedState) LastJustifiedSlot() uint64 {
+	return c.data.LastJustifiedSlot
+}
+
+// LastFinalizedSlot returns the last finalized Slot of the beacon chain.
+func (c *CrystallizedState) LastFinalizedSlot() uint64 {
+	return c.data.LastFinalizedSlot
+}
+
+// CurrentDynasty returns the current dynasty of the beacon chain.
+func (c *CrystallizedState) CurrentDynasty() uint64 {
+	return c.data.CurrentDynasty
+}
+
+// TotalDeposits returns total balance of deposits.
+func (c *CrystallizedState) TotalDeposits() uint64 {
+	return c.data.TotalDeposits
+}
+
+// DynastySeed is used to select the committee for each shard.
+func (c *CrystallizedState) DynastySeed() [32]byte {
+	var h [32]byte
+	copy(h[:], c.data.DynastySeed)
+	return h
+}
+
+// Validators returns list of validators.
+func (c *CrystallizedState) Validators() []*pb.ValidatorRecord {
+	return c.data.Validators
+}
+
+// IsCycleTransition checks if a new cycle has been reached. At that point,
+// a new crystallized state and active state transition will occur.
+func (c *CrystallizedState) IsCycleTransition(slotNumber uint64) bool {
+	return slotNumber >= c.LastStateRecalc()+params.CycleLength
+}
+
+// getAttesterIndices fetches the attesters for a given attestation record.
+func (c *CrystallizedState) getAttesterIndices(attestation *pb.AttestationRecord) ([]uint32, error) {
+	slotsStart := int64(c.LastStateRecalc()) - params.CycleLength
+	slotIndex := int64(attestation.Slot) - slotsStart
+	// TODO: ShardAndCommitteesForSlots will return default value because the spec for dynasty transition is not finalized.
+	shardCommitteeArray := c.data.ShardAndCommitteesForSlots
+	shardCommittee := shardCommitteeArray[slotIndex].ArrayShardAndCommittee
+	for i := 0; i < len(shardCommittee); i++ {
+		if attestation.ShardId == shardCommittee[i].ShardId {
+			return shardCommittee[i].Committee, nil
+		}
+	}
+	return nil, fmt.Errorf("unable to find attestation based on slot: %v, shardID: %v", attestation.Slot, attestation.ShardId)
+}
+
+// CalculateNewCrystallizedState computes the new crystallized state, given the previous crystallized state
+// and the current active state. This method is called during a cycle transition.
+func (c *CrystallizedState) CalculateNewCrystallizedState(aState *ActiveState, slotNumber uint64) (*CrystallizedState, error) {
+	var blockVoteBalance uint64
+	justifiedStreak := c.JustifiedStreak()
+	justifiedSlot := c.LastJustifiedSlot()
+	finalizedSlot := c.LastFinalizedSlot()
+	lastStateRecalc := c.LastStateRecalc()
+	blockVoteCache := aState.GetBlockVoteCache()
+
+	// walk through all the slots from LastStateRecalc - cycleLength to LastStateRecalc - 1.
```

Comment:
I don't understand this comment and how it relates to the code that follows it. Is there anything we can add or improve to this comment to make the next few lines easier to understand?

This loops from 0 to CycleLength(64), not from LastStateRecalc - cycleLength to LastStateRecalc - 1.

Reply:
I think it would help if the comments also included "and determine if each block is justified by using the balance stored in the vote cache". I've looked at this `blockVoteCache` logic before and I'm not convinced that it's correct, and was planning on taking another look after this PR.

But for now, this is copy/pasta of old code.